unmon_standard10.pkl > array code

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


## 1. 데이터 불러오기

In [ ]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt

# X와 y 데이터 불러오기
file_path_1 = '/content/drive/My Drive/dataset/feature/data_arrays_1.npz'
file_path_2 = '/content/drive/My Drive/dataset/feature/data_arrays_2.npz'
file_path_3 = '/content/drive/My Drive/dataset/feature/data_arrays_3.npz'
file_path_4 = '/content/drive/My Drive/dataset/feature/data_arrays_4.npz'
file_path_5 = '/content/drive/My Drive/dataset/feature/unmon_data_1.npz'
file_path_6 = '/content/drive/My Drive/dataset/feature/unmon_data_2.npz'

# npz 파일 읽기 (allow_pickle=True 옵션 추가)
data_1 = np.load(file_path_1, allow_pickle=True)
data_2 = np.load(file_path_2, allow_pickle=True)
data_3 = np.load(file_path_3, allow_pickle=True)
data_4 = np.load(file_path_4, allow_pickle=True)

#이 밑에서부터는 unmon
data_5 = np.load(file_path_5, allow_pickle=True) #X1, X5, X6, X7, X9, X12, X14의 정보를 담고 있다
data_6 = np.load(file_path_6, allow_pickle=True) #X3, X5, X8, X11, X15, X116, X17의 정보를 담고 있다

# X 데이터 추출
# y = data_1['y']
y_mon = data_1['y']
X1_mon = data_1['X1']
X3_mon = data_1['X3']
X5_mon = data_2['X5']  # X1 데이터 로드
X6_mon = data_2['X6']
X7_mon = data_2['X7']
X8_mon = data_3['X8']
X9_mon = data_3['X9']
X11_mon = data_3['X11']
X12_mon = data_3['X12']  # X1 데이터 로드
X14_mon = data_4['X14']
X15_mon = data_4['X15']
X16_mon = data_4['X16']
X17_mon = data_4['X17']

X1_unmon = data_5['X1_unmon']
X5_unmon = data_5['X5_unmon']  # X1 데이터 로드
X6_unmon = data_5['X6_unmon']
X7_unmon = data_5['X7_unmon']
X9_unmon = data_5['X9_unmon']
X12_unmon = data_5['X12_unmon']  # X1 데이터 로드
X14_unmon = data_5['X14_unmon']

X3_unmon = data_6['X3']
X8_unmon = data_6['X8']
X11_unmon = data_6['X11']
X15_unmon = data_6['X15']
X16_unmon = data_6['X16']
X17_unmon = data_6['X17']

data_mon = pd.DataFrame({
    'Feature1': X1_mon,
    'Feature3': X3_mon,
    'Feature5': X5_mon,
    'Feature6': X6_mon,
    'Feature7': X7_mon,
    'Feature8': X8_mon,
    'Feature9': X9_mon,
    #'Feature11': X11_mon,
    'Feature12': X12_mon,
    'Feature14': X14_mon,
    #'Feature15': X15_mon,
    'Feature16': X16_mon,
    #'Feature17': X17_mon
})

data_unmon = pd.DataFrame({
    'Feature1': X1_unmon,
    'Feature3': X3_unmon,
    'Feature5': X5_unmon,
    'Feature6': X6_unmon,
    'Feature7': X7_unmon,
    'Feature8': X8_unmon,
    'Feature9': X9_unmon,
    #'Feature11': X11_unmon, #중요도 검사후 주석 처리 함
    'Feature12': X12_unmon,
    'Feature14': X14_unmon,
    #'Feature15': X15_unmon, #중요도 검사후 주석 처리 함
    'Feature16': X16_unmon,
    #'Feature17': X17_unmon  #중요도 검사후 주석 처리 함
})

In [ ]:
df_mon = pd.DataFrame(data_mon)
df_mon

Feature1  Feature3  Feature5  Feature6  Feature7  Feature8  Feature9  \
0      0.007141   26060.8      1300  0.085151  0.914849  1.934185       121   
1      0.019652   14796.8       438  0.154440  0.845560  1.540626        80   
2      0.008187   24729.6      1240  0.086892  0.913108  1.973276       118   
3      0.009246   27136.0      1324  0.084371  0.915629  2.832813       122   
4      0.007573   24422.4      1291  0.081792  0.918208  2.179823       115   
...         ...       ...       ...       ...       ...       ...       ...   
18995  0.004655   36812.8      8815  0.065614  0.934386  4.342246       619   
18996  0.001567   41216.0      9404  0.055444  0.944556  2.758278       552   
18997  0.001500   34252.8      9373  0.058179  0.941821  2.893279       579   
18998  0.002006   37683.2      9236  0.069514  0.930486  3.313856       690   
18999  0.001387   36505.6      9168  0.076272  0.923728  2.720712       757   

       Feature12  Feature14  Feature16  
0           2842       1421   1.018253  
1           1036        518   1.328057  
2           2716       1358   1.068847  
3           2892       1446   1.556907  
4           2812       1406   1.180701  
...          ...        ...        ...  
18995      18868       9434   3.029885  
18996      19912       9956   2.665505  
18997      19904       9952   2.613620  
18998      19852       9926   2.998973  
18999      19850       9925   2.191058  

[19000 rows x 10 columns]

In [ ]:
df_unmon = pd.DataFrame(data_unmon)
df_unmon

Feature1  Feature3  Feature5  Feature6  Feature7  Feature8  Feature9  \
0     0.043101     16384       101  0.223077  0.776923  1.050736        29   
1     0.001197     51200      9189  0.074343  0.925657  1.940547       738   
2     0.049749     16896       285  0.206128  0.793872  3.475381        74   
3     0.000922     43520      9407  0.054383  0.945617  1.313723       541   
4     0.021376     35840      2215  0.116826  0.883174  8.507451       293   
...        ...       ...       ...       ...       ...       ...       ...   
9995  0.006988     35328      4180  0.089919  0.910081  5.127064       413   
9996  0.007559     44544      4663  0.087476  0.912524  4.348233       447   
9997  0.097028     20992       302  0.163435  0.836565  7.707609        59   
9998  0.023307     21504       413  0.188605  0.811395  2.842843        96   
9999  0.000963     46080      9668  0.032232  0.967768  1.295610       322   

      Feature12  Feature14  Feature16  
0           260        130   1.240463  
1         19854       9927   1.572821  
2           718        359   2.939175  
3         19896       9948   1.016587  
4          5016       2508   5.034849  
...         ...        ...        ...  
9995       9186       4593   5.704984  
9996      10220       5110   3.668209  
9997        722        361   0.868159  
9998       1018        509   2.130534  
9999      19980       9990   0.997824  

[10000 rows x 10 columns]

In [ ]:
df_y_mon = pd.DataFrame(y_mon)
df_y_mon

0
0       0
1       0
2       0
3       0
4       0
...    ..
18995  94
18996  94
18997  94
18998  94
18999  94

[19000 rows x 1 columns]

##unmon 데이터프레임으로 합치기

In [ ]:
y_unmon = {'0': [-1] * 10000}
df_y_unmon = pd.DataFrame(y_unmon)
df_y_unmon

0
0    -1
1    -1
2    -1
3    -1
4    -1
...  ..
9995 -1
9996 -1
9997 -1
9998 -1
9999 -1

[10000 rows x 1 columns]

##2. unmon + mon 합쳐서 total 데이터셋 만들기

In [ ]:
data_x_total = pd.DataFrame({
    'Feature1': [X1_unmon, X1_mon],
    'Feature3': [X3_unmon, X3_mon],  # 추가할 Feature3
    'Feature5': [X5_unmon, X5_mon],
    'Feature6': [X6_unmon, X6_mon],
    'Feature7': [X7_unmon, X7_mon],
    'Feature8': [X8_unmon, X8_mon],
    'Feature9': [X9_unmon, X9_mon],
    #'Feature11': [X11_unmon, X11_mon], #마찬가지로 중요도 검사 후 주석 처리함.
    'Feature12': [X12_unmon, X12_mon],
    'Feature14': [X14_unmon, X14_mon],
    #'Feature15': [X15_unmon, X15_mon], #마찬가지로 중요도 검사 후 주석 처리함.
    'Feature16': [X16_unmon, X16_mon],
    #'Feature17': [X17_unmon, X17_mon] #마찬가지로 중요도 검사 후 주석 처리함.
})

data_total = pd.concat([df_unmon, df_mon], ignore_index=True)

In [ ]:
data_total

Feature1  Feature3  Feature5  Feature6  Feature7  Feature8  Feature9  \
0      0.043101   16384.0       101  0.223077  0.776923  1.050736        29   
1      0.001197   51200.0      9189  0.074343  0.925657  1.940547       738   
2      0.049749   16896.0       285  0.206128  0.793872  3.475381        74   
3      0.000922   43520.0      9407  0.054383  0.945617  1.313723       541   
4      0.021376   35840.0      2215  0.116826  0.883174  8.507451       293   
...         ...       ...       ...       ...       ...       ...       ...   
28995  0.004655   36812.8      8815  0.065614  0.934386  4.342246       619   
28996  0.001567   41216.0      9404  0.055444  0.944556  2.758278       552   
28997  0.001500   34252.8      9373  0.058179  0.941821  2.893279       579   
28998  0.002006   37683.2      9236  0.069514  0.930486  3.313856       690   
28999  0.001387   36505.6      9168  0.076272  0.923728  2.720712       757   

       Feature12  Feature14  Feature16  
0            260        130   1.240463  
1          19854       9927   1.572821  
2            718        359   2.939175  
3          19896       9948   1.016587  
4           5016       2508   5.034849  
...          ...        ...        ...  
28995      18868       9434   3.029885  
28996      19912       9956   2.665505  
28997      19904       9952   2.613620  
28998      19852       9926   2.998973  
28999      19850       9925   2.191058  

[29000 rows x 10 columns]

In [ ]:
data_y_total = pd.concat([df_y_unmon.iloc[:,0], df_y_mon.iloc[:,0]], axis=0, ignore_index=True)
data_y_total = data_y_total.to_frame()  # Series를 DataFrame으로 변환

In [ ]:
data_y_total

0
0      -1
1      -1
2      -1
3      -1
4      -1
...    ..
28995  94
28996  94
28997  94
28998  94
28999  94

[29000 rows x 1 columns]

#공통 data split 용.

In [ ]:
from sklearn.model_selection import train_test_split

# 데이터 분할
X_train, X_test, y_train, y_test = train_test_split(data_total, data_y_total, test_size=0.2, random_state=42)

## 모델 Training

#1.logistic regression
 (1.데이터샘플링,2.차원축소, 3.미니배치)

#1) Logistic Regression Accuracy (Sampled Data): 0.3431034482758621

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(

    해당 오류는 결과값이 선형으로 수렴하지 않는다는 경고이다. 즉, 우리의 데이터는 복잡한 비선형 구조를 띄고 있다고 이해할 수 있다.

In [ ]:
#1. 데이터 샘플링:
#무작위로 데이터를 샘플링하여 작은 규모의 훈련 데이터셋을 만들고 Logistic Regression 모델을 훈련하는 코드입니다.

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

# 데이터 샘플링
sampled_data = data_total.sample(frac=0.1, random_state=42)
sampled_labels = data_y_total.loc[sampled_data.index]

# 데이터 분할
X_train_sampled, X_test_sampled, y_train_sampled, y_test_sampled = train_test_split(
    sampled_data, sampled_labels, test_size=0.2, random_state=42
)


# Logistic Regression 모델 훈련
clf_LR_sampled = LogisticRegression(max_iter=5000)
clf_LR_sampled.fit(X_train_sampled, y_train_sampled.values.ravel())

# 모델 평가
y_pred_sampled = clf_LR_sampled.predict(X_test_sampled)
accuracy_sampled = accuracy_score(y_test_sampled, y_pred_sampled)
classification_report_sampled = classification_report(y_test_sampled, y_pred_sampled, zero_division=1)

print("Logistic Regression Accuracy (Sampled Data):", accuracy_sampled)
print("Logistic Regression Classification Report (Sampled Data):")
print(classification_report_sampled)

Logistic Regression Accuracy (Sampled Data): 0.3431034482758621
Logistic Regression Classification Report (Sampled Data):
              precision    recall  f1-score   support

          -1       0.36      0.99      0.52       194
           0       1.00      0.00      0.00         4
           1       1.00      0.00      0.00         4
           2       1.00      0.00      0.00         4
           3       1.00      0.00      0.00         5
           4       1.00      0.00      0.00         5
           5       1.00      0.00      0.00         6
           6       1.00      0.00      0.00         5
           7       1.00      0.00      0.00         3
           8       1.00      0.00      0.00         8
           9       1.00      0.00      0.00         9
          10       1.00      0.00      0.00         7
          11       1.00      0.00      0.00         5
          12       1.00      0.00      0.00         2
          13       1.00      0.00      0.00         1
          14 

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


결과값을 해석하자면 -1로 할당된 class가 너무 많다. 즉, 클래스가 불균형한 것이다. 이런 경우 튜닝이 필요하다.

In [ ]:
#1. data sampling 튜닝하기
from imblearn.over_sampling import RandomOverSampler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report
from sklearn.model_selection import train_test_split

# 데이터 샘플링 (Random Oversampling)
oversample = RandomOverSampler(sampling_strategy='auto', random_state=42)
X_over, y_over = oversample.fit_resample(X_train_sampled, y_train_sampled)

# 클래스 -1에 대한 가중치는 1로 설정
class_weights = {-1: 1}

# 나머지 클래스에 대한 가중치 설정
for i in range(94):
    if i != -1:
        class_weights[i] = 5 # 클래스 -1의 샘플 수에 비례하여 가중치 설정 (10,000 / 2000)


# Logistic Regression 모델 생성 시 class_weight 매개변수를 지정하여 가중치 부여
clf = LogisticRegression(class_weight=class_weights, max_iter=5000)

# 모델 학습
clf.fit(X_over, y_over)

# 모델 평가
y_pred = clf.predict(X_test_sampled)
accuracy = accuracy_score(y_test_sampled, y_pred)
classification_report_output = classification_report(y_test_sampled, y_pred, zero_division=1)

print("Logistic Regression Accuracy:", accuracy)
print("Logistic Regression Classification Report:")
print(classification_report_output)

/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


Logistic Regression Accuracy: 0.056896551724137934
Logistic Regression Classification Report:
              precision    recall  f1-score   support

          -1       1.00      0.00      0.00       194
           0       0.00      0.00      0.00         4
           1       0.00      0.00      0.00         4
           2       1.00      0.00      0.00         4
           3       1.00      0.00      0.00         5
           4       1.00      0.00      0.00         5
           5       0.09      0.17      0.12         6
           6       0.00      0.00      0.00         5
           7       0.04      0.33      0.08         3
           8       1.00      0.00      0.00         8
           9       1.00      0.00      0.00         9
          10       1.00      0.00      0.00         7
          11       0.50      0.20      0.29         5
          12       0.00      0.00      0.00         2
          13       1.00      0.00      0.00         1
          14       1.00      0.00      0.

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


#2)Logistic Regression Accuracy (PCA): 0.34413793103448276


In [ ]:
from sklearn.decomposition import PCA
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report
from sklearn.model_selection import train_test_split

# 데이터를 훈련 및 테스트 세트로 나누기
X_train, X_test, y_train, y_test = train_test_split(data_total, data_y_total, test_size=0.2, random_state=1)

# 차원 축소 모델 생성 및 훈련 데이터에 적용
pca = PCA(n_components=7)
X_train_pca = pca.fit_transform(X_train)
X_test_pca = pca.transform(X_test)

# Logistic Regression 모델 훈련
clf_LR_pca = LogisticRegression(max_iter=5000)
clf_LR_pca.fit(X_train_pca, y_train.values.ravel())

# 모델 평가
y_pred_pca = clf_LR_pca.predict(X_test_pca)
accuracy_pca = accuracy_score(y_test, y_pred_pca)
classification_report_pca = classification_report(y_test, y_pred_pca)

print("Logistic Regression Accuracy (PCA):", accuracy_pca)
print("Logistic Regression Classification Report (PCA):")
print(classification_report_pca)

Logistic Regression Accuracy (PCA): 0.32
Logistic Regression Classification Report (PCA):
              precision    recall  f1-score   support

          -1       0.56      0.77      0.65      2058
           0       0.00      0.00      0.00        41
           1       0.00      0.00      0.00        35
           2       0.00      0.00      0.00        43
           3       0.00      0.00      0.00        41
           4       0.00      0.00      0.00        44
           5       0.00      0.00      0.00        40
           6       0.00      0.00      0.00        31
           7       0.00      0.00      0.00        42
           8       0.00      0.00      0.00        43
           9       0.00      0.00      0.00        37
          10       0.00      0.00      0.00        43
          11       0.06      0.30      0.09        30
          12       0.00      0.00      0.00        42
          13       0.00      0.00      0.00        37
          14       0.00      0.00      0.00  

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels wit

#3)Logistic Regression Accuracy (Batch Training): 0.31517241379310346

In [ ]:
#3.배치 훈련:
#배치 크기를 지정하여 모델을 여러 에폭동안 훈련하는 코드입니다.
from sklearn.linear_model import SGDClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report
import numpy as np

# 데이터 분할
X_train_batch, X_test_batch, y_train_batch, y_test_batch = train_test_split(
    data_total, data_y_total, test_size=0.2, random_state=42
)

# Logistic Regression 모델 훈련 (배치 훈련)
batch_size = 1000
num_epochs = 10

clf_LR_batch = SGDClassifier(max_iter=1000, random_state=42)

for epoch in range(num_epochs):
    # 무작위로 배치 생성
    indices = np.random.choice(len(X_train_batch), batch_size, replace=False)
    X_batch = X_train_batch.iloc[indices]
    y_batch = y_train_batch.iloc[indices].values.ravel()

    # 모델 훈련
    clf_LR_batch.partial_fit(X_batch, y_batch, classes=np.unique(y_train_batch))

# 모델 평가
y_pred_batch = clf_LR_batch.predict(X_test_batch)
accuracy_batch = accuracy_score(y_test_batch, y_pred_batch)
classification_report_batch = classification_report(y_test_batch, y_pred_batch)

print("Logistic Regression Accuracy (Batch Training):", accuracy_batch)

print("Logistic Regression Classification Report (Batch Training):")
print(classification_report_batch)


Logistic Regression Accuracy (Batch Training): 0.006379310344827587
Logistic Regression Classification Report (Batch Training):
              precision    recall  f1-score   support

          -1       0.00      0.00      0.00      1979
           0       0.00      0.00      0.00        42
           1       0.00      0.00      0.00        50
           2       0.00      0.00      0.00        43
           3       0.00      0.00      0.00        30
           4       0.00      0.00      0.00        47
           5       0.00      0.00      0.00        39
           6       0.00      0.00      0.00        40
           7       0.00      0.00      0.00        33
           8       0.00      0.00      0.00        43
           9       0.00      0.00      0.00        44
          10       0.00      0.00      0.00        45
          11       0.00      0.00      0.00        46
          12       0.00      0.00      0.00        49
          13       0.00      0.00      0.00        43
       

/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


# 2. Decision tree

#Decision Tree Accuracy (in %): 59.95%

In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report

# 데이터 분할
X_train, X_test, y_train, y_test = train_test_split(data_total, data_y_total, test_size=0.2, random_state=42)

# 결정 트리 모델 생성 및 훈련
clf_decision_tree = DecisionTreeClassifier(random_state=42)
clf_decision_tree.fit(X_train, y_train)

# 모델 평가
y_pred_decision_tree = clf_decision_tree.predict(X_test)
accuracy_decision_tree = accuracy_score(y_test, y_pred_decision_tree)
classification_report_decision_tree = classification_report(y_test, y_pred_decision_tree)

print("Decision Tree Accuracy (in %): {:.2f}%".format(accuracy_decision_tree * 100))

print("Decision Tree Classification Report:")
print(classification_report_decision_tree)


Decision Tree Accuracy (in %): 59.69%
Decision Tree Classification Report:
              precision    recall  f1-score   support

          -1       0.82      0.83      0.82      1979
           0       0.32      0.29      0.30        42
           1       0.26      0.16      0.20        50
           2       0.65      0.60      0.63        43
           3       0.46      0.53      0.49        30
           4       0.46      0.34      0.39        47
           5       0.41      0.36      0.38        39
           6       0.64      0.70      0.67        40
           7       0.41      0.52      0.46        33
           8       0.59      0.51      0.55        43
           9       0.47      0.41      0.44        44
          10       0.40      0.44      0.42        45
          11       0.51      0.54      0.53        46
          12       0.83      0.69      0.76        49
          13       0.46      0.37      0.41        43
          14       0.30      0.35      0.32        31
      

# 튜닝한 Decision Tree
1)
Best parameters: {'max_depth': None, 'min_samples_leaf': 3, 'min_samples_split': 2}

2)
Accuracy of the best model: 0.6017241379310345


In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV

# Decision Tree 모델과 관련된 하이퍼파라미터 그리드 설정
param_grid = {
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 3, 4],
    'min_samples_leaf': [1, 2, 3]
}

# GridSearchCV를 사용하여 최적의 하이퍼파라미터 찾기
grid_search = GridSearchCV(DecisionTreeClassifier(random_state=42), param_grid, cv=5)
grid_search.fit(X_train, y_train)

# 최적의 하이퍼파라미터 출력
print("Best parameters:", grid_search.best_params_)

# 최적의 모델로 평가
tree_best_model = grid_search.best_estimator_
tree_accuracy_best_model = tree_best_model.score(X_test, y_test)
print("Accuracy of the best model:", tree_accuracy_best_model)



Best parameters: {'max_depth': None, 'min_samples_leaf': 1, 'min_samples_split': 2}
Accuracy of the best model: 0.5968965517241379


In [ ]:
from sklearn.metrics import classification_report

# 최적의 모델로 예측
tree_y_pred = tree_best_model.predict(X_test)

# 최적의 모델 평가 및 classification_report 출력
tree_classification_report = classification_report(y_test, tree_y_pred)
print("Decision Tree Classification Report for the Best Model:")
print(tree_classification_report)

Decision Tree Classification Report for the Best Model:
              precision    recall  f1-score   support

          -1       0.82      0.83      0.82      1979
           0       0.32      0.29      0.30        42
           1       0.26      0.16      0.20        50
           2       0.65      0.60      0.63        43
           3       0.46      0.53      0.49        30
           4       0.46      0.34      0.39        47
           5       0.41      0.36      0.38        39
           6       0.64      0.70      0.67        40
           7       0.41      0.52      0.46        33
           8       0.59      0.51      0.55        43
           9       0.47      0.41      0.44        44
          10       0.40      0.44      0.42        45
          11       0.51      0.54      0.53        46
          12       0.83      0.69      0.76        49
          13       0.46      0.37      0.41        43
          14       0.30      0.35      0.32        31
          15       0.48  

## 3. Random Forest

#Random Forest Accuracy (in %): 71.16%

In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report

# 랜덤 포레스트 모델 생성 및 훈련
clf_random_forest = RandomForestClassifier(random_state=42)
clf_random_forest.fit(X_train, y_train.values.ravel())

# 모델 평가
y_pred_random_forest = clf_random_forest.predict(X_test)
accuracy_random_forest = accuracy_score(y_test, y_pred_random_forest)
classification_report_random_forest = classification_report(y_test, y_pred_random_forest)

print("Random Forest Accuracy (in %): {:.2f}%".format(accuracy_random_forest * 100))

print("Random Forest Classification Report:")
print(classification_report_random_forest)


Random Forest Accuracy (in %): 71.16%
Random Forest Classification Report:
              precision    recall  f1-score   support

          -1       0.78      0.92      0.84      1979
           0       0.67      0.43      0.52        42
           1       0.50      0.26      0.34        50
           2       0.71      0.79      0.75        43
           3       0.67      0.60      0.63        30
           4       0.60      0.53      0.56        47
           5       0.68      0.59      0.63        39
           6       0.77      0.75      0.76        40
           7       0.62      0.70      0.66        33
           8       0.70      0.70      0.70        43
           9       0.63      0.55      0.59        44
          10       0.82      0.51      0.63        45
          11       0.74      0.57      0.64        46
          12       0.81      0.80      0.80        49
          13       0.56      0.35      0.43        43
          14       0.52      0.48      0.50        31
      

원래는 binary와 같이 feature 7개로 진행했으나 결과가 좋지 않았다. 13개의 feature 전부 사용하면 과적합 문제가 생길 것 같아서 중요도 검사를 따로 진행했다.

In [ ]:
feature_importance = clf_random_forest.feature_importances_
num_features = len(feature_importance)

# 특성 중요도 출력
for i in range(num_features):
    print(f"Feature {i + 1} importance: {feature_importance[i]}")

Feature 1 importance: 0.08014748879310223
Feature 2 importance: 0.17460484298053447
Feature 3 importance: 0.08203407857980052
Feature 4 importance: 0.0759151780112168
Feature 5 importance: 0.07633207567829499
Feature 6 importance: 0.06585863956517138
Feature 7 importance: 0.07470483673401732
Feature 8 importance: 0.048872228406383146
Feature 9 importance: 0.08217429328035099
Feature 10 importance: 0.08101040093739358
Feature 11 importance: 0.03574390690169854
Feature 12 importance: 0.06888110287652849
Feature 13 importance: 0.053720927255507575


In [ ]:
from sklearn.ensemble import RandomForestClassifier

# RandomForestClassifier 모델 생성 (criterion='entropy'로 설정)
clf_random_forest_entropy = RandomForestClassifier(random_state=42, criterion='entropy')
clf_random_forest_entropy.fit(X_train, y_train.ravel())

# 특성 중요도 (Entropy) 출력
feature_importance_entropy = clf_random_forest_entropy.feature_importances_

for i, importance in enumerate(feature_importance_entropy):
    print(f"Feature {i + 1} importance (Entropy): {importance}")

Feature 1 importance (Entropy): 0.07624400386397329
Feature 2 importance (Entropy): 0.12832974956999255
Feature 3 importance (Entropy): 0.10210637065563802
Feature 4 importance (Entropy): 0.09671005313940702
Feature 5 importance (Entropy): 0.09622189408277418
Feature 6 importance (Entropy): 0.049700345505096305
Feature 7 importance (Entropy): 0.10017825051594402
Feature 8 importance (Entropy): 0.029310696669663655
Feature 9 importance (Entropy): 0.10014307429015896
Feature 10 importance (Entropy): 0.09935673059834187
Feature 11 importance (Entropy): 0.04166101092730532
Feature 12 importance (Entropy): 0.04726137795485427
Feature 13 importance (Entropy): 0.03277644222685048


In [ ]:
# 각 특성별 Gini 중요도와 Entropy 중요도를 리스트로 표현
gini_importance = [
    0.08014748879310223, 0.17460484298053447, 0.08203407857980052,
    0.0759151780112168, 0.07633207567829499, 0.06585863956517138,
    0.07470483673401732, 0.048872228406383146, 0.08217429328035099,
    0.08101040093739358, 0.03574390690169854, 0.06888110287652849,
    0.053720927255507575
]

entropy_importance = [
    0.07624400386397329, 0.12832974956999255, 0.10210637065563802,
    0.09671005313940702, 0.09622189408277418, 0.049700345505096305,
    0.10017825051594402, 0.029310696669663655, 0.10014307429015896,
    0.09935673059834187, 0.04166101092730532, 0.04726137795485427,
    0.03277644222685048
]

# 각 특성별 Gini와 Entropy 중요도 합산
combined_importance = [gini + entropy for gini, entropy in zip(gini_importance, entropy_importance)]

# 각 특성별 평균 중요도 계산
avg_combined_importance = [importance / 2 for importance in combined_importance]

# 결과 출력
for idx, importance in enumerate(avg_combined_importance):
    print(f"Feature {idx + 1} importance (Combined Gini and Entropy): {importance}")# 각 특성별 Gini 중요도와 Entropy 중요도를 리스트로 표현
gini_importance = [
    0.08014748879310223, 0.17460484298053447, 0.08203407857980052,
    0.0759151780112168, 0.07633207567829499, 0.06585863956517138,
    0.07470483673401732, 0.048872228406383146, 0.08217429328035099,
    0.08101040093739358, 0.03574390690169854, 0.06888110287652849,
    0.053720927255507575
]

entropy_importance = [
    0.07624400386397329, 0.12832974956999255, 0.10210637065563802,
    0.09671005313940702, 0.09622189408277418, 0.049700345505096305,
    0.10017825051594402, 0.029310696669663655, 0.10014307429015896,
    0.09935673059834187, 0.04166101092730532, 0.04726137795485427,
    0.03277644222685048
]

# 각 특성별 Gini와 Entropy 중요도 합산
combined_importance = [gini + entropy for gini, entropy in zip(gini_importance, entropy_importance)]

# 각 특성별 평균 중요도 계산
avg_combined_importance = [importance / 2 for importance in combined_importance]


# 결과 출력
for idx, importance in enumerate(avg_combined_importance):
    print(f"Feature {idx + 1} importance (Combined Gini and Entropy): {importance}")

Feature 1 importance (Combined Gini and Entropy): 0.07819574632853776
Feature 2 importance (Combined Gini and Entropy): 0.1514672962752635
Feature 3 importance (Combined Gini and Entropy): 0.09207022461771927
Feature 4 importance (Combined Gini and Entropy): 0.08631261557531192
Feature 5 importance (Combined Gini and Entropy): 0.08627698488053459
Feature 6 importance (Combined Gini and Entropy): 0.05777949253513384
Feature 7 importance (Combined Gini and Entropy): 0.08744154362498066
Feature 8 importance (Combined Gini and Entropy): 0.0390914625380234
Feature 9 importance (Combined Gini and Entropy): 0.09115868378525498
Feature 10 importance (Combined Gini and Entropy): 0.09018356576786772
Feature 11 importance (Combined Gini and Entropy): 0.03870245891450193
Feature 12 importance (Combined Gini and Entropy): 0.05807124041569138
Feature 13 importance (Combined Gini and Entropy): 0.04324868474117903
Feature 1 importance (Combined Gini and Entropy): 0.07819574632853776
Feature 2 importan


Feature 1 importance (Combined Gini and Entropy): 0.07819574632853776

Feature 2 importance (Combined Gini and Entropy): 0.1514672962752635

Feature 3 importance (Combined Gini and Entropy): 0.09207022461771927

Feature 4 importance (Combined Gini and Entropy): 0.08631261557531192

Feature 5 importance (Combined Gini and Entropy): 0.08627698488053459

Feature 6 importance (Combined Gini and Entropy): 0.05777949253513384

Feature 7 importance (Combined Gini and Entropy): 0.08744154362498066

Feature 8 importance (Combined Gini and Entropy): 0.0390914625380234

Feature 9 importance (Combined Gini and Entropy): 0.09115868378525498

Feature 10 importance (Combined Gini and Entropy): 0.09018356576786772

Feature 11 importance (Combined Gini and Entropy): 0.03870245891450193

Feature 12 importance (Combined Gini and Entropy): 0.05807124041569138

Feature 13 importance (Combined Gini and Entropy): 0.04324868474117903

위에서 0.05 이하인 것들을 추리면 X8, X11, X13 정도다. 즉 feature 1. 3. 5. 6. 7. 8. 9. 12. 14. 16만 남겨두면 된다.  

다시 처음으로 돌아가서 주석 처리하고 다시 돌려보겠다.


피쳐 처리 전 : Random Forest Accuracy (in %): 72.31%
피쳐 처리 후 : 71.16

#튜닝 : Accuracy of the best model: 0.7222413793103448

Best parameters: {'max_depth': 28, 'max_features': 'auto', 'min_samples_leaf': 1, 'min_samples_split': 4, 'n_estimators': 766}

In [ ]:
from scipy.stats import randint
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import RandomizedSearchCV

# 튜닝할 하이퍼파라미터 범위 설정
param_dist = {
    'n_estimators': randint(400, 800),  # 100에서 1000 사이의 랜덤한 정수
    'max_depth': [None] + list(randint(10, 30).rvs(5)),  # None과 10에서 30 사이의 랜덤한 정수
    'min_samples_split': randint(3, 15),  # 2에서 20 사이의 랜덤한 정수
    'min_samples_leaf': randint(1, 5),   # 1에서 10 사이의 랜덤한 정수
    'max_features': ['sqrt', 'auto']
}

# RandomForestClassifier 생성
clf = RandomForestClassifier(random_state=42)

# RandomizedSearchCV를 사용하여 튜닝
random_search = RandomizedSearchCV(clf, param_distributions=param_dist, n_iter=50, cv=5, random_state=42)
random_search.fit(X_train, y_train.values.ravel())

# 최적의 하이퍼파라미터 출력
print("Best parameters:", random_search.best_params_)

# 최적의 모델로 평가
best_model = random_search.best_estimator_
accuracy_best_model = best_model.score(X_test, y_test)
print("Accuracy of the best model:", accuracy_best_model)

#이거 오래걸림...

/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_forest.py:424: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomForestClassifiers and ExtraTreesClassifiers.
  warn(
/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_forest.py:424: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomForestClassifiers and ExtraTreesClassifiers.
  warn(
/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_forest.py:424: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomFor

Best parameters: {'max_depth': 28, 'max_features': 'auto', 'min_samples_leaf': 1, 'min_samples_split': 4, 'n_estimators': 766}
Accuracy of the best model: 0.7222413793103448


In [ ]:
from sklearn.metrics import classification_report

# 최적의 모델로 예측
y_pred = best_model.predict(X_test)

# classification_report 출력
report = classification_report(y_test, y_pred)
print("Classification Report:")
print(report)

Classification Report:
              precision    recall  f1-score   support

          -1       0.79      0.92      0.85      2058
           0       0.66      0.56      0.61        41
           1       0.65      0.49      0.56        35
           2       0.64      0.79      0.71        43
           3       0.77      0.56      0.65        41
           4       0.63      0.50      0.56        44
           5       0.73      0.60      0.66        40
           6       0.72      0.90      0.80        31
           7       0.72      0.62      0.67        42
           8       0.81      0.91      0.86        43
           9       0.61      0.54      0.57        37
          10       0.63      0.44      0.52        43
          11       0.49      0.57      0.52        30
          12       0.81      0.71      0.76        42
          13       0.62      0.41      0.49        37
          14       0.58      0.51      0.55        35
          15       0.61      0.70      0.65        33
    

In [ ]:
from sklearn.ensemble import RandomForestClassifier

# 최적의 하이퍼파라미터로 모델 생성
best_model = RandomForestClassifier(max_depth=28, max_features='auto', min_samples_leaf=2, min_samples_split=4, n_estimators=100, random_state=42)

# 모델 학습
best_model.fit(X_train, y_train.values.ravel())

# 테스트 데이터에 대해 모델 평가
accuracy_best_model = best_model.score(X_test, y_test)
print("Accuracy of the best model:", accuracy_best_model)

/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_forest.py:424: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomForestClassifiers and ExtraTreesClassifiers.
  warn(


Accuracy of the best model: 0.713103448275862


In [ ]:
from sklearn.metrics import classification_report

# 모델 예측
y_pred = best_model.predict(X_test)

# classification_report를 이용하여 precision, recall, f1-score 등 확인
report = classification_report(y_test, y_pred)
print("Classification Report:")
print(report)

Classification Report:
              precision    recall  f1-score   support

          -1       0.78      0.92      0.84      1979
           0       0.65      0.48      0.55        42
           1       0.58      0.28      0.38        50
           2       0.70      0.77      0.73        43
           3       0.62      0.60      0.61        30
           4       0.61      0.57      0.59        47
           5       0.72      0.54      0.62        39
           6       0.76      0.78      0.77        40
           7       0.62      0.70      0.66        33
           8       0.68      0.70      0.69        43
           9       0.62      0.57      0.60        44
          10       0.81      0.49      0.61        45
          11       0.69      0.54      0.61        46
          12       0.83      0.78      0.80        49
          13       0.58      0.33      0.42        43
          14       0.50      0.42      0.46        31
          15       0.77      0.73      0.75        51
    

## 4. k-NN

#KNN Accuracy (in %): 44.95%

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, classification_report

# KNN 모델 생성 및 훈련
clf_knn = KNeighborsClassifier()
clf_knn.fit(X_train, y_train)

# y_train, y_test를 1차원 배열로 변환
y_train = np.ravel(y_train)
y_test = np.ravel(y_test)

# 변환 후 KNN 모델 훈련 및 예측
clf_knn.fit(X_train, y_train)
y_pred_knn = clf_knn.predict(X_test)

# 모델 평가
y_pred_knn = clf_knn.predict(X_test)
accuracy_knn = accuracy_score(y_test, y_pred_knn)
classification_report_knn = classification_report(y_test, y_pred_knn)

print("KNN Accuracy (in %): {:.2f}%".format(accuracy_knn * 100))


/usr/local/lib/python3.10/dist-packages/sklearn/neighbors/_classification.py:215: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)


KNN Accuracy (in %): 44.95%


#튜닝한 K-NN

1)
Best K value: 8

Tuned KNN Accuracy (in %): 46.48%

In [ ]:
from sklearn.model_selection import GridSearchCV

# 탐색할 이웃 수의 범위 지정
param_grid = {'n_neighbors': range(1, 21)}

# 그리드 서치를 사용하여 최적의 이웃 수 찾기
grid_search = GridSearchCV(KNeighborsClassifier(), param_grid, cv=5)
grid_search.fit(X_train, y_train)

# 최적의 이웃 수 출력
print("Best K value:", grid_search.best_params_['n_neighbors'])

# 최적의 이웃 수로 모델 재훈련
best_k = grid_search.best_params_['n_neighbors']
clf_knn_tuned = KNeighborsClassifier(n_neighbors=best_k)
clf_knn_tuned.fit(X_train, y_train)

# 튜닝된 모델 평가
y_pred_knn_tuned = clf_knn_tuned.predict(X_test)
accuracy_knn_tuned = accuracy_score(y_test, y_pred_knn_tuned)
classification_report_knn_tuned = classification_report(y_test, y_pred_knn_tuned)

print("Tuned KNN Accuracy (in %): {:.2f}%".format(accuracy_knn_tuned * 100))
print("Classification Report for Tuned KNN:\n", classification_report_knn_tuned)

/usr/local/lib/python3.10/dist-packages/sklearn/neighbors/_classification.py:215: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.10/dist-packages/sklearn/neighbors/_classification.py:215: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.10/dist-packages/sklearn/neighbors/_classification.py:215: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.10/dist-packages/sklearn/neighbors/_classification.py:215: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for 

Best K value: 8
Tuned KNN Accuracy (in %): 46.48%
Classification Report for Tuned KNN:
               precision    recall  f1-score   support

          -1       0.69      0.92      0.79      2058
           0       0.09      0.07      0.08        41
           1       0.11      0.14      0.12        35
           2       0.26      0.49      0.34        43
           3       0.33      0.37      0.35        41
           4       0.36      0.23      0.28        44
           5       0.07      0.07      0.07        40
           6       0.04      0.06      0.05        31
           7       0.22      0.26      0.24        42
           8       0.34      0.37      0.36        43
           9       0.12      0.14      0.12        37
          10       0.16      0.07      0.10        43
          11       0.11      0.17      0.13        30
          12       0.54      0.50      0.52        42
          13       0.20      0.22      0.21        37
          14       0.27      0.40      0.33    

##5. SVM (rbf, linear)

#SVM Accuracy (in %): 34.19%


In [ ]:
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report

# SVM 모델 생성 및 훈련
svm_classifier = SVC(kernel='rbf', C=1.0, gamma='scale')  # rbf 커널 사용, C와 gamma 값은 예시입니다.
svm_classifier.fit(X_train, y_train)

# 모델 평가
y_pred_svm = svm_classifier.predict(X_test)
accuracy_svm = accuracy_score(y_test, y_pred_svm)
classification_report_svm = classification_report(y_test, y_pred_svm)

print("SVM Accuracy (in %): {:.2f}%".format(accuracy_svm * 100))
print("Classification Report for SVM:\n", classification_report_svm)

/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


SVM Accuracy (in %): 34.19%
Classification Report for SVM:
               precision    recall  f1-score   support

          -1       0.35      1.00      0.51      1979
           0       0.00      0.00      0.00        42
           1       0.00      0.00      0.00        50
           2       0.00      0.00      0.00        43
           3       0.00      0.00      0.00        30
           4       0.00      0.00      0.00        47
           5       0.00      0.00      0.00        39
           6       0.00      0.00      0.00        40
           7       0.00      0.00      0.00        33
           8       0.00      0.00      0.00        43
           9       0.00      0.00      0.00        44
          10       0.00      0.00      0.00        45
          11       0.00      0.00      0.00        46
          12       0.00      0.00      0.00        49
          13       0.00      0.00      0.00        43
          14       0.00      0.00      0.00        31
          15       0.

/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


#튜닝한 RGF SVM : 65.86%
1)Best parameters: {'C': 100, 'gamma': 1}

2)Tuned SVM Accuracy (in %): 65.86%

In [ ]:
from sklearn.svm import SVC
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import accuracy_score

# 데이터 스케일링
scaler = StandardScaler()

y_train = y_train.ravel()
y_test = y_test.values.ravel()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)


# SVM 모델 생성
svm_classifier = SVC()

# 탐색할 매개변수 그리드 설정
param_grid = {
    'C': [0.1, 1, 10, 100],
    'gamma': ['scale', 'auto', 0.1, 1, 10]
}

# 그리드 서치를 사용하여 최적의 매개변수 찾기
grid_search = GridSearchCV(svm_classifier, param_grid, cv=5)
grid_search.fit(X_train_scaled, y_train)

# 최적의 매개변수 출력
print("Best parameters:", grid_search.best_params_)

# 최적의 매개변수로 모델 재훈련
best_C = grid_search.best_params_['C']
best_gamma = grid_search.best_params_['gamma']
svm_classifier_tuned = SVC(C=best_C, gamma=best_gamma)
svm_classifier_tuned.fit(X_train_scaled, y_train)

# 튜닝된 모델 평가
y_pred_tuned = svm_classifier_tuned.predict(X_test_scaled)
accuracy_tuned = accuracy_score(y_test, y_pred_tuned)
print("Tuned SVM Accuracy (in %): {:.2f}%".format(accuracy_tuned * 100))

In [ ]:
from sklearn.svm import SVC
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score

# 데이터 스케일링
scaler = StandardScaler()

y_train = y_train.values.ravel()
y_test = y_test.values.ravel()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# 최적의 하이퍼파라미터로 새로운 SVM 모델 생성
best_C = 100
best_gamma = 1
svm_classifier_tuned = SVC(C=best_C, gamma=best_gamma)
svm_classifier_tuned.fit(X_train_scaled, y_train)

# 튜닝된 모델 평가
y_pred_tuned = svm_classifier_tuned.predict(X_test_scaled)
accuracy_tuned = accuracy_score(y_test, y_pred_tuned)
print("Tuned SVM Accuracy (in %): {:.2f}%".format(accuracy_tuned * 100))

Tuned SVM Accuracy (in %): 68.31%


In [ ]:
y_pred_tuned = svm_classifier_tuned.predict(X_test_scaled)
accuracy_tuned = accuracy_score(y_test, y_pred_tuned)
print("Tuned SVM Accuracy (in %): {:.2f}%".format(accuracy_tuned * 100))

# Classification Report 출력
print("Classification Report:")
print(classification_report(y_test, y_pred_tuned))

Tuned SVM Accuracy (in %): 68.31%
Classification Report:
              precision    recall  f1-score   support

          -1       0.77      0.88      0.82      2058
           0       0.59      0.49      0.53        41
           1       0.43      0.43      0.43        35
           2       0.60      0.72      0.65        43
           3       0.70      0.46      0.56        41
           4       0.72      0.64      0.67        44
           5       0.64      0.45      0.53        40
           6       0.73      0.87      0.79        31
           7       0.69      0.79      0.73        42
           8       0.76      0.74      0.75        43
           9       0.56      0.51      0.54        37
          10       0.72      0.42      0.53        43
          11       0.48      0.53      0.51        30
          12       0.79      0.79      0.79        42
          13       0.34      0.35      0.35        37
          14       0.52      0.43      0.47        35
          15       0.50 

##2) Linear SVM: 0.11137931034482759


In [ ]:
from sklearn.svm import SVC, LinearSVC
from sklearn.metrics import accuracy_score

# 선형 커널 SVM 모델 학습 및 평가
svm_linear = LinearSVC(C=10000, loss='hinge')
svm_linear.fit(X_train, y_train)
y_pred_linear = svm_linear.predict(X_test)
accuracy_linear = accuracy_score(y_test, y_pred_linear)
print("Accuracy of Linear SVM:", accuracy_linear)

/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


Accuracy of Linear SVM: 0.011206896551724138


/usr/local/lib/python3.10/dist-packages/sklearn/svm/_base.py:1244: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


#튜닝한 Linear SVM

1) y = column_or_1d(y, warn=True)

Best Parameters: {'C': 1, 'loss': 'hinge'}

Best Cross-Validation Accuracy: 0.14439655172413796

Test Set Accuracy of Best Model: 0.010689655172413793

In [ ]:
from sklearn.model_selection import GridSearchCV

# 탐색할 매개변수 그리드 정의
param_grid = {'C': [0.1, 1, 10, 100, 1000],
              'loss': ['hinge', 'squared_hinge']}

# SVC 모델 객체 생성
svm = LinearSVC()

# GridSearchCV 객체 생성 및 훈련 데이터에 대해 그리드 탐색 수행
grid_search = GridSearchCV(svm, param_grid, cv=5, scoring='accuracy')
grid_search.fit(X_train, y_train)

# 최적의 하이퍼파라미터와 그 때의 정확도 출력
print("Best Parameters:", grid_search.best_params_)
print("Best Cross-Validation Accuracy:", grid_search.best_score_)

# 최적의 모델에 대해 테스트 세트로 성능 평가
best_svm = grid_search.best_estimator_
test_accuracy = best_svm.score(X_test, y_test)
print("Test Set Accuracy of Best Model:", test_accuracy)

/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.10/dist-packages/sklearn/svm/_base.py:1244: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.10/dist-packages/sklearn/svm/_base.py:1244: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d 

Best Parameters: {'C': 1, 'loss': 'hinge'}
Best Cross-Validation Accuracy: 0.14439655172413796
Test Set Accuracy of Best Model: 0.010689655172413793


/usr/local/lib/python3.10/dist-packages/sklearn/svm/_base.py:1244: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
